Make multiple window work

josealvim opened the pull request:
I made the new code follow a more friendly standard, besides making multiple windows work

Commit:
make desktop::manager let us poke at its windows

Commit:
Add a vector of render screens to the graphic::manager

ugdk compiles but we havent tested the examples

Commit:
abort

Commit:
merge the abort

Commit:
begin path to multiple windows

Commit:
fix a merge

Commit:
fix a merge

Commit:
I just realized I wasn't working on ugdk-ex../externals/ugdk

Commit:
attempt to change ugdk to make multiwindowing work

Commit:
Final correction to graphic and desktop

I think

Right now the problem with blank window is
input handling, when the cursor moves into
the blank window, or starts moving inside
the window, the engine seg-faults, which
indicates a problem with the sdl handling
of inputs

Commit:
Add a feature to get windows by their SDL id

Commit:
Make Blank window work

	Okay, what happened is we implemented a get by id method
	which allows the sdl-input things work without segfaulting
	all over the place

Commit:
make apropriate changes to make multi window work

Commit:
Make Multiple windows WOOOORK!

Commit:
remove debugging couts

Commit:
Start reworking the examples to make them not look horrible

	Multiple windows made things ugly, now we are working
	to make it _not ugly_

	As of now, I'm changing the code so that we have multiple
	render functions, which now take only one canvas const &
	and the engine:Run() will call all the render functions in
	order to write on their window and present them

Commit:
Work around lambda madness

Commit:
make using multiple windows dev friendly

Kazuo256 commented on modules/ugdk-core/include/ugdk/desktop/manager.h:
```diff
@@ -26,17 +27,21 @@ class Manager {
     std::weak_ptr<Window> primary_window_;
-    std::map< uint32, std::shared_ptr<Window> > windows_;
+    std::vector< std::shared_ptr<Window> > windows_;
+    std::unordered_map<uint32_t, uint32_t> map_;
```
Use a clearer name for `map_`. What does it map?

Kazuo256 commented on modules/ugdk-core/include/ugdk/graphic/manager.h:
```diff
@@ -34,11 +34,20 @@ class Manager {
+    bool Initialize(
+        const std::vector<std::weak_ptr<desktop::Window>>& windows_, 
+        const math::Vector2D& canvas_size
+                   );
```
This formatting does not follow the standard. The appropriate would be:
```c++
bool Initialize(const std::vector<std::weak_ptr<desktop::Window>>& windows_, 
                const math::Vector2D& canvas_size);
```

Kazuo256 commented on modules/ugdk-core/include/ugdk/graphic/manager.h:
```diff
@@ -34,11 +34,20 @@ class Manager {
+
+    void UseCanvas(graphic::Canvas &);
+
+    //void AttachCanvasToScreen(uint32_t index, const std::weak_ptr<desktop::Window>& weak_window);
```
What is this?

josealvim replied:
leftover code

Kazuo256 commented on modules/ugdk-core/src/action/scene.cc:
```diff
@@ -45,14 +48,28 @@ void Scene::Focus() {
+}
+void Scene::set_render_function(uint32_t index
+                        , const std::function<void (graphic::Canvas&)>& function
+                        ) {
```
Again, bad formatting.

Kazuo256 commented on modules/ugdk-core/src/graphic/canvas.cc:
```diff
@@ -26,18 +26,18 @@ Canvas::Canvas(RenderTarget* render_target)
     }
     ACTIVE_CANVAS = this;
-    Bind();
+    Bind();*/
```
If this is no longer necessary, then remove it.

Kazuo256 commented on modules/ugdk-core/src/graphic/canvas.cc:
```diff
@@ -51,6 +51,7 @@ Canvas::~Canvas() {
         next_canvas_->previous_canvas_ = previous_canvas_;
     if (previous_canvas_)
         previous_canvas_->next_canvas_ = next_canvas_;
+    */
```
If this is no longer necessary, then remove it.

Kazuo256 commented on modules/ugdk-core/src/graphic/manager.cc:
```diff
@@ -77,13 +98,18 @@ void Manager::SetUserNearestNeighborTextures(bool enabled) {
+bool Manager::Initialize(
+            const std::vector<std::weak_ptr<desktop::Window>>& windows_, 
+            const math::Vector2D& canvas_size
+    ) {
```
Bad formatting

josealvim replied:
Corrected

Kazuo256 commented on modules/ugdk-core/src/graphic/manager.cc:
```diff
@@ -92,10 +118,11 @@ bool Manager::Initialize(const std::weak_ptr<desktop::Window>& window_weak, cons
+        //screens_[i].reset(new RenderScreen);
+        ResizeScreen(i, canvas_size);
+    }
+    //SetActiveScreen(0);
```
Remove commented lines if they are no longer necessary

Kazuo256 commented on modules/ugdk-core/src/graphic/manager.cc:
```diff
@@ -138,7 +165,7 @@ bool Manager::Initialize(const std::weak_ptr<desktop::Window>& window_weak, cons
 void Manager::Release() {
     SDL_GL_DeleteContext(context_);
-    screen_.reset();
+    screens_[0].reset();
```
Why only the zero-th screen?

Kazuo256 commented on modules/ugdk-core/src/graphic/module.cc:
```diff
@@ -1,13 +1,19 @@
+bool Initialize(std::unique_ptr<Manager> manager, 
+                const std::weak_ptr<desktop::Window> & window,
+                const math::Vector2D &canvas_size
+                ) {
```
Bad formatting (this was close though)

Kazuo256 commented on modules/ugdk-core/src/graphic/rendertarget.cc:
```diff
@@ -32,13 +34,13 @@ void RenderTarget::Clear(structure::Color color) {
 
 void RenderTarget::Bind() {
-    system::AssertCondition<system::InvalidOperation>(!IsActive(), "RenderTarget must be inactive for Bind.");
+    //system::AssertCondition<system::InvalidOperation>(!IsActive(), "RenderTarget must be inactive for Bind.");
```
Are you sure it's ok to remove this assertion?

josealvim replied:
Yes, nowhere else does the code check for the render target being active

Kazuo256 commented on modules/ugdk-core/src/graphic/rendertarget.cc:
```diff
@@ -32,13 +34,13 @@ void RenderTarget::Clear(structure::Color color) {
 
 void RenderTarget::Unbind() {
-    system::AssertCondition<system::InvalidOperation>(IsActive(), "RenderTarget must be active for Unbind.");
+    //system::AssertCondition<system::InvalidOperation>(IsActive(), "RenderTarget must be active for Unbind.");
```
Are you sure it's ok to remove this assertion?

josealvim replied:
Yes, nowhere else does the code, or the examples for that matter, check or rely on it being active

josealvim replied:
It's okay to unbind things that are unbound, because when ~Canvas is called, it tries to unbind things.
I kept the assert on Bind() so things are still safe, the assert on Unbind isn't necessary.

Kazuo256 replied:
Then just remove the line

Kazuo256 commented on modules/ugdk-core/src/system/engine.cc:
```diff
@@ -232,11 +233,31 @@ void Run() {
+                auto &manager = graphic::manager();
+                std::vector<graphic::Canvas*> canvases;
+                
+                //std::cout << manager.num_screens() << " Screens" << std::endl;
```
Remove print code!

Kazuo256 commented on modules/ugdk-core/src/system/engine.cc:
```diff
@@ -232,11 +233,31 @@ void Run() {
+
+                            desktop::manager().window(i).lock()->Present();//show or lose it
+                        }
+                //desktop::manager().PresentAll();
```
Remove line if it is no longer necessary

Commit:
Abide to change requests

Commit:
Conform to PR reviews and make map_id_to_index more reasonable

Commit:
make ugdk able to process mutiple windows

we still need to change the close window event however

Commit:
Make UGDK clean up it's screens when closing windows

It doesn't track all windows always...
some segfaults when closing windows in
the middle of the array...

Commit:
Make it work

Commit:
conform to code style on manager.cc

Kazuo256 commented on modules/ugdk-core/include/ugdk/action/scene.h:
```diff
@@ -82,10 +82,10 @@ class Scene : public system::TaskPlayer {
+    
+    uint32_t num_functions();
+    void set_render_function(uint32_t, const std::function<void (graphic::Canvas&)>&);
+    void remove_render_function(uint32_t);
```
Write the parameter names: they help documenting the function, since this is the source file Doxygen uses to generate the API reference

Kazuo256 commented on modules/ugdk-core/include/ugdk/graphic/manager.h:
```diff
@@ -33,12 +33,20 @@ class Manager {
-    void AttachTo(const std::shared_ptr<desktop::Window>&);
-    void ResizeScreen(const math::Vector2D& canvas_size);
+    void RegisterScreen(std::weak_ptr<desktop::Window>);
+    void DeregisterScreen(uint32_t index);
```
**Un**register

Kazuo256 commented on modules/ugdk-core/src/action/scene.cc:
```diff
@@ -45,14 +48,31 @@ void Scene::Focus() {
+        render_functions_[index] = function;
+    /* Check if there are enough render-targets ? */
+}
+void Scene::remove_render_function(uint32_t index) {
```
This is neither a getter nor a setter, so the name formatting should be PascalCase.

josealvim replied:
We have a little problem with the example now

Kazuo256 commented on modules/ugdk-core/src/system/engine.cc:
```diff
@@ -232,11 +233,40 @@ void Run() {
+                auto &manager = graphic::manager();
+                std::vector<graphic::Canvas*> canvases;
+                std::vector<uint32_t> kill_these_screens;
+                //std::cout << manager.num_screens() << " Screens" << std::endl;
```
Remove commented line

Kazuo256 commented on modules/ugdk-core/src/system/engine.cc:
```diff
@@ -232,11 +233,40 @@ void Run() {
+                                scene->Render(i, *canvases[i]);//do it
+
+                                if (desktop::manager().window(i).lock())
+                                    desktop::manager().window(i).lock()->Present();//show or lose it
```
Please use a temporaray variable

Commit:
Conform to PR changes and add a dirty temporary hot fix

josealvim commented:
More problems:
The problem doesn't lie in
- desktop manager.cc, 
I've tested it and it only deletes the correct windows
- dynamic-windows.cc
It couldn't be from here either

josealvim commented:
Okay, problem lies on engine.cc:258

Somehow, when I exit out of a window, the last window's pointer gets invalidated, something to do with the desktop::manager

Commit:
Change code so we won't double delete windows

Commit:
Place vector declaration in the right scope, so it will DIE

Commit:
re-indent a line VSCode screwed up
